specs/ChainSpec/Consensus.md

```diff
@@ -0,0 +1,146 @@
+
+```python
+def on_block(self, block):
+    header = block.header
+
+    if header.height <= self.head_height:
+        return
+
+    last_final_block = store.get_block(header.last_final_block_hash)
+
+    self.head_height = header.height
+    self.head_hash = block.hash()
+    self.largest_final_height = last_final_block.height
+
+    self.timer_height = self.head_height + 1
+    self.timer_started = time.time()
+
+    self.endorsement_pending = True
+```
+
+The timer needs to be checked periodically, and contain the following logic:
+
+```python
+def get_delay(n):
+    min(MAX_DELAY, MIN_DELAY + DELAY_STEP * (n-2))
+
+def process_timer(self):
+    now = time.time()
+
+    skip_delay = get_delay(self.timer_height - self.largest_final_height)
+
+    if self.endorsement_pending and now > self.timer_started + ENDORSEMENT_DELAY:
+
+        if self.head_height >= self.largest_target_height:
+            self.larget_target_height = self.head_height + 1
+            self.send_approval(head_height + 1)
+
+        self.endorsement_pending = False
+
+    if now > self.timer_started + skip_delay:
+        assert not self.endorsement_pending
+
+        self.largest_target_height = max(self.largest_target_height, self.timer_height + 1)
+        self.send_approval(self.timer_height + 1)
+
+        self.timer_started = now
+        self.timer_height += 1
+```
+
+Where `send_approval` creates an approval with `prev_hash` and `prev_height` equal to `self.head_hash` and `self.head_height` correspondingly, `target_height` to the passed argument, and sends to the block producer at `target_height`.
+
+## Block Production
+We first define a convenience function to fetch approvals that can be included in a block at particular height:
+
+```python
+def get_approvals(self, target_height):
+    return [approval for approval
+                     in self.approvals
+                     if approval.target_height == target_height and
+                        approval.prev_height == self.head_height and
+                        (approval.prev_height + 1 != target_height or
+                         approval.prev_hash == self.head_hash)]
+```
+
+A block producer assigned for a particular height produces a block at that height whenever they have `get_approvals` return approvals from block producers whose stake collectively exceeds 2/3 of the total stake.
+
+## Safety
+
+We call approvals for which `prev_height + 1 = target_height` endorsements, and all other approvals skips.
+
+Note that with the implementation above a honest block producer can never produce two endorsements with the same `prev_height` (call this condition *conflicting endorsements*), neither can they produce a skip message `s` and an endorsement `e` such that `s.prev_height > e.prev_height and s.target_height <= e.target_height` (call this condition *conflicting skip and endorsement*).
```

Comment:
Did you mean `s.prev_height <= e.prev_height and s.target_height >= e.target_height`?